proposals/BVAL-508.md

```diff
@@ -0,0 +1,257 @@
+Containers can contain one value, or several values.
+Extractors for specific containers can be implemented.
+
+TODO: should we differentiate single and plural?
+
+    // Draft proposal for the contract
+    public interface SingleContainerValueExtractor<CONTAINER,CONTAINED> {
+        CONTAINED extractValue(CONTAINER container);
+    }
+    
+    public interface ManyContainerValuesExtractor<CONTAINER,CONTAINED> {
+        Iterable<CONTAINED> extractValue(CONTAINER container);
+    }
+    
+    // Examples
+    
+    //TODO can T remain unspecified? or should it be ? or Object?
+    public class OptionalContainerExtractor<T> implements SingleContainerValueExtractor<Optional<T>,T> {
+        T extractValue(Optional<T> optional) { ... };
+    }
+
+Bean Validation will have out of the box support for containers for `Iterable` and `Map`.
+
+TODO: any other Java SE type to consider?
+TODO: should we support JavaFX Property<T>?
+
+Custom container extractors can be provided via:
+
+* property
+* factory
+* service locator
+
+TODO: refine how the container implementations are discovered.
+See [Hibernate Validator's feature](http://docs.jboss.org/hibernate/validator/5.2/reference/en-US/html_single/#section-value-handling) 
+
+#### When are containers used?
+
+When a property is of a known container, constraints declarations are explored.
+A typical example is a property using constraints on type parameters of `TYPE_USE` (`Collection<@Size String>`).
+
+If constraints are discovered, the unwrapping mechanism is used
+to read contained value(s) and potentially apply constraints on them (see below).
+
+By constraints, we also mean property traversal `@Valid`.
+
+Containers of containers are applied recursively if necessary.
+TODO: infinite loops?
+
+#### Which container unwrapper to use?
+
+The container used is the one corresponding to the most specific super type
+as defined in the BV resolution of constraint validators.
+
+### Constraints applied and containers
+
+Constraints declared on the type parameter of a type use will be applied to the contained value
+as extracted by the container logic.
+
+    // each String of the collection is validated for the regexp
+    Set<@Pattern(...) String> emails;
+
+By default, constraints declared on the container will apply to the container.
+This ensures backward compatibility.
+
+    // @Size is applied on the collection
+    @Size(min=5) List<Integer> ages;
+
+Extractors can specify that constraints declared on the container apply to the contained value(s);
+
+    @ConstraintsApplyTo(CONTAINED_VALUES)
+    public class JavaFXPropertyContainerExtractor<T> implements SingleContainerValueExtractor<Property<T>,T> { ... }
+    
+    // test that the age is at least 5
+    @Min(5) IntegerProperty<Integer> age;
+
+This is useful for JavaFX to force the validation of the contained properties.
+
+One can also force the constraints to apply to the container or the container value per site
```

Comment:
> Should that be a defined attribute of constraint annotations instead? 

Right. In HV we were forced to use additional annotations, but as part of the spec the constraint annotation itself could change.